File: tests/c#/10/VersionTests.cs
using Xunit;

namespace Hardstuck.GuildWars2.BuildCodes.V2.Tests.Version;

public class VersionTests {
	public static IEnumerable<object[]> DataProvider() {
		var invalidCodes = new[] {
			TestUtilities.CodesInvalid["wrong-version"],
		};
		
		var v1Codes = TestUtilities.CodesV1.Values;
		
		var v2Codes = TestUtilities.CodesV2.Values;

		return invalidCodes.Select(c => new object[] { c, -1 })
			.Concat(v1Codes.Select(c => new object[] { c,  1 }))
			.Concat(v2Codes.Select(c => {
				var version = TextLoader.INVERSE_CHARSET[c[0]];
				if(version >= 26) version -= 26;
				return new object[] { c, version };
			}));
	}

	[Theory] [MemberData(nameof(DataProvider))]
	public void VersionDetectionTest(string code, int expectedVersion) {
		Assert.Equal(expectedVersion, Static.DetermineCodeVersion(code));
	}
}

File: tests/php/8.0/VersionTests.php
<?php namespace Hardstuck\GuildWars2\BuildCodes\V2\Tests\API;

require_once 'TestUtilities.php';

use Hardstuck\GuildWars2\BuildCodes\V2\Statics;
use Hardstuck\GuildWars2\BuildCodes\V2\Tests\TestUtilities;
use Hardstuck\GuildWars2\BuildCodes\V2\TextLoader;
use PHPUnit\Framework\TestCase;

class VersionTests extends TestCase {
	public function DataProvider() {
		$invalidCodes = [
			TestUtilities::$CodesInvalid["wrong-version"],
		];
		
		$v1Codes = TestUtilities::$CodesV1;
		
		$v2Codes = TestUtilities::$CodesV2;

		return array_merge(
			array_map(fn($code) => [$code, -1], $invalidCodes),
			array_map(fn($code) => [$code, 1], $v1Codes),
			array_map(function($code) {
				$version = TextLoader::INVERSE_CHARSET[ord($code[0])];
				if($version >= 26) $version -= 26;
				return [$code, $version];
			}, $v2Codes),
		 );
	}

	/** @test @dataProvider DataProvider */
	public function VersionDetectionTest(string $code, int $expectedVersion) {
		$this->assertEquals($expectedVersion, Statics::DetermineCodeVersion($code));
	}
}
